File: lua/protein-translation/protein-translation.lua
local codons = {
	AUG = "Methionine",
	UUU = "Phenylalanine",
	UUC = "Phenylalanine",
	UUA = "Leucine",
	UUG = "Leucine",
	UCU = "Serine",
	UCC = "Serine",
	UCA = "Serine",
	UCG = "Serine",
	UAU = "Tyrosine",
	UAC = "Tyrosine",
	UGU = "Cysteine",
	UGC = "Cysteine",
	UGG = "Tryptophan",
	UAA = "STOP",
	UAG = "STOP",
	UGA = "STOP",
}
local function translate_codon(codon)
	return codons[codon] or error()
end

local function translate_rna_strand(rna_strand)
	local startIndx = 1
	local translation = {}
	while startIndx < #rna_strand do
		local codon = string.sub(rna_strand, startIndx, startIndx + 2)
		local protein = codons[codon]
		assert(protein)
		if protein == "STOP" then
			return translation
		end
		table.insert(translation, protein)
		startIndx = startIndx + 3
	end
	return translation
end

return {
	codon = translate_codon,
	rna_strand = translate_rna_strand,
}

File: typescript/protein-translation/protein-translation.ts
function codons(rnaArr: string[], output: string[] = []): string[] {
  const [first, second, third, ...rest] = rnaArr
  const chunked = [...output, first + second + third]
  return rest.length > 0 ? codons(rest, chunked) : chunked
}

const proteinMap: Record<string, string> = {
  AUG: 'Methionine',
  UUU: 'Phenylalanine',
  UUC: 'Phenylalanine',
  UUA: 'Leucine',
  UUG: 'Leucine',
  UCU: 'Serine',
  UCC: 'Serine',
  UCA: 'Serine',
  UCG: 'Serine',
  UAU: 'Tyrosine',
  UAC: 'Tyrosine',
  UGU: 'Cysteine',
  UGC: 'Cysteine',
  UGG: 'Tryptophan',
  UAA: 'STOP',
  UAG: 'STOP',
  UGA: 'STOP',
}

export function translate(rna: string): string[] {
  const proteins = codons([...rna]).map((codon) => proteinMap[codon])
  const output = []
  for (const protein of proteins) {
    if (!protein) throw new Error('Invalid codon')
    if (protein === 'STOP') return output
    output.push(protein)
  }
  return output
}
